File: euler/src/problems/Problem3.cs
using System;
using System.Linq;
using System.Collections.Generic;

namespace euler {

    class Problem3 {
        public static void LargestPrimeFactor(long n) {
            Console.WriteLine(primes(n, 2, new HashSet<long>()).Max());
        }

        private static ISet<long> primes(long n, long p, ISet<long> found) {
            if (n < p * p) {
                found.Add(n);
                return found;
            }
            else if (n % p == 0) {
                found.Add(p);
                return primes(n / p, p, found);
            }
            else {
                return primes(n, p + 1, found);
            }
        }
    }
}

File: euler/src/problems/Problem3.scala
package problems

import scala.annotation.tailrec
import scala.collection.mutable

object Problem3 extends App {

  println(primeFactors(600851475143L))
  println(primeFactors(600851475143L).max)

  def primeFactors(n: Long): Set[Long] = {
    @tailrec
    def primes(n: Long, p: Long = 2L, found: mutable.Builder[Long, Set[Long]] = Set.newBuilder[Long]): Set[Long] = {
      if (n < p * p)
        (found += n).result()
      else if (n % p == 0)
        primes(n / p, p, found += p)
      else
        primes(n, p + 1L, found)
    }
    
    primes(n)
  }
}
